llvm/lib/Transforms/Coroutines/Coroutines.cpp

```diff
@@ -456,52 +482,70 @@ static void addCallToCallGraph(CallGraph *CG, CallInst *Call, Function *Callee){
 
 Value *coro::Shape::emitAlloc(IRBuilder<> &Builder, Value *Size,
                               CallGraph *CG) const {
+  unsigned sizeParamIndex = UINT_MAX;
   switch (ABI) {
   case coro::ABI::Switch:
     llvm_unreachable("can't allocate memory in coro switch-lowering");
 
   case coro::ABI::Retcon:
-  case coro::ABI::RetconOnce: {
-    auto Alloc = RetconLowering.Alloc;
-    Size = Builder.CreateIntCast(Size,
-                                 Alloc->getFunctionType()->getParamType(0),
-                                 /*is signed*/ false);
-    ConstantInt* TypeId = RetconLowering.TypeId;
-    CallInst *Call;
-    if (TypeId == nullptr)
-      Call = Builder.CreateCall(Alloc, Size);
-    else
-      Call = Builder.CreateCall(Alloc, {Size, TypeId});
-    propagateCallAttrsFromCallee(Call, Alloc);
-    addCallToCallGraph(CG, Call, Alloc);
-    return Call;
-  }
+  case coro::ABI::RetconOnce:
+    sizeParamIndex = 0;
+    break;
+  case coro::ABI::RetconOnceDynamic:
+    sizeParamIndex = 1;
+    break;
   case coro::ABI::Async:
     llvm_unreachable("can't allocate memory in coro async-lowering");
   }
-  llvm_unreachable("Unknown coro::ABI enum");
+  auto Alloc = RetconLowering.Alloc;
+  Size = Builder.CreateIntCast(
+      Size, Alloc->getFunctionType()->getParamType(sizeParamIndex),
+      /*is signed*/ false);
+  SmallVector<Value *, 2> Args;
+  if (ABI == coro::ABI::RetconOnceDynamic) {
+    Args.push_back(RetconLowering.Allocator);
+  }
+  Args.push_back(Size);
+  if (ABI == coro::ABI::RetconOnce) {
+    ConstantInt *TypeId = RetconLowering.TypeId;
+    if (TypeId != nullptr)
+      Args.push_back(TypeId);
+  }
+  auto *Call = Builder.CreateCall(Alloc, Args);
+  propagateCallAttrsFromCallee(Call, Alloc);
+  addCallToCallGraph(CG, Call, Alloc);
+  return Call;
 }
 
 void coro::Shape::emitDealloc(IRBuilder<> &Builder, Value *Ptr,
                               CallGraph *CG) const {
+  unsigned sizeParamIndex = UINT_MAX;
   switch (ABI) {
   case coro::ABI::Switch:
     llvm_unreachable("can't allocate memory in coro switch-lowering");
 
   case coro::ABI::Retcon:
-  case coro::ABI::RetconOnce: {
-    auto Dealloc = RetconLowering.Dealloc;
-    Ptr = Builder.CreateBitCast(Ptr,
-                                Dealloc->getFunctionType()->getParamType(0));
-    auto *Call = Builder.CreateCall(Dealloc, Ptr);
-    propagateCallAttrsFromCallee(Call, Dealloc);
-    addCallToCallGraph(CG, Call, Dealloc);
-    return;
-  }
+  case coro::ABI::RetconOnce:
+    sizeParamIndex = 0;
+    break;
+  case coro::ABI::RetconOnceDynamic:
+    sizeParamIndex = 1;
+    break;
   case coro::ABI::Async:
     llvm_unreachable("can't allocate memory in coro async-lowering");
   }
-  llvm_unreachable("Unknown coro::ABI enum");
+  auto Dealloc = RetconLowering.Dealloc;
```

Comment:
Do we consistently check that the switch over ABI is covered. If so I would not regress that.

Reply:
Fixed.